File: rlm/cli/state/pid.py
"""Helpers de PID — fonte única (DRY) para todo o CLI."""

from __future__ import annotations

import os
import subprocess
import sys
from pathlib import Path


def read_pid_file(pid_path: Path) -> int | None:
    """Lê e retorna o PID de *pid_path*, ou ``None`` se inválido/ausente."""
    try:
        return int(pid_path.read_text(encoding="utf-8").strip())
    except Exception:
        return None


def pid_alive(pid: int) -> bool:
    """Retorna True se o processo *pid* está em execução.

    Windows: usa ``tasklist``; POSIX: ``os.kill(pid, 0)``.
    """
    if sys.platform == "win32":
        try:
            result = subprocess.run(
                ["tasklist", "/fi", f"PID eq {pid}", "/nh", "/fo", "csv"],
                capture_output=True,
                text=True,
                timeout=5,
            )
            return f'"{pid}"' in result.stdout or str(pid) in result.stdout
        except Exception:
            return False
    try:
        os.kill(pid, 0)
        return True
    except (ProcessLookupError, PermissionError):
        return False


def write_pid(pid_file: Path, pid: int) -> None:
    """Grava *pid* em *pid_file*, criando diretórios necessários."""
    pid_file.parent.mkdir(parents=True, exist_ok=True)
    pid_file.write_text(str(pid))


def remove_pid(pid_file: Path) -> None:
    """Remove *pid_file* silenciosamente."""
    pid_file.unlink(missing_ok=True)

File: packages/cli/src/state/pid.ts
/**
 * state/pid.ts — Gerenciamento de arquivos PID.
 *
 * Porta fiel de rlm/cli/state/pid.py
 */

import { readFileSync, writeFileSync, unlinkSync, mkdirSync } from "node:fs";
import { execSync } from "node:child_process";
import { dirname } from "node:path";

/** Lê o PID de um arquivo, retorna null se inválido/ausente. */
export function readPidFile(pidPath: string): number | null {
  try {
    const text = readFileSync(pidPath, "utf8").trim();
    const n = parseInt(text, 10);
    return Number.isFinite(n) && n > 0 ? n : null;
  } catch {
    return null;
  }
}

/** Retorna true se o processo com o PID dado está em execução. */
export function pidAlive(pid: number): boolean {
  if (process.platform === "win32") {
    try {
      const out = execSync(
        `tasklist /fi "PID eq ${pid}" /nh /fo csv`,
        { encoding: "utf8", timeout: 5000 }
      );
      return out.includes(`"${pid}"`) || out.includes(String(pid));
    } catch {
      return false;
    }
  }
  // POSIX: kill -0 não mata, apenas verifica existência
  try {
    process.kill(pid, 0);
    return true;
  } catch (err: unknown) {
    const code = (err as NodeJS.ErrnoException).code;
    // EPERM = processo existe mas sem permissão → está vivo
    return code === "EPERM";
  }
}

/** Grava o PID no arquivo, criando diretórios necessários. */
export function writePid(pidFile: string, pid: number): void {
  mkdirSync(dirname(pidFile), { recursive: true });
  writeFileSync(pidFile, String(pid), "utf8");
}

/** Remove o arquivo PID silenciosamente. */
export function removePid(pidFile: string): void {
  try {
    unlinkSync(pidFile);
  } catch {
    // ignora se não existe
  }
}
